### PR TITLE
Improve type handling in loading and particle execution

### PR DESCRIPTION
**Pull request by lindner**
- Add documention to Loader to help describe the steps taken.
- Add types for load/require Particle, adjust callers and tests
- Add types to slot-dom-consumer, particle-execution-context etc.
- Identified a problematic issue with readonly properties.

### CODE REVIEW

**Comment by lindner** [comment]
Note that this is stacked on top of  #2916

File: src/runtime/particle-execution-context.ts [hunk 1]
```diff
@@ -195,19 +195,24 @@ export class ParticleExecutionContext {
...
+  async _instantiateParticle(id: string, spec: ParticleSpec, proxies: Map<string, StorageProxy>): Promise<[any, () => Promise<void>]> {
     let resolve : () => void = null;
     const p = new Promise<void>(res => resolve = res);
     this.pendingLoads.push(p);
     const clazz = await this.loader.loadParticleClass(spec);
     const capabilities = this.defaultCapabilitySet();
     const particle = new clazz(); // TODO: how can i add an argument to DomParticle ctor?
-    particle.id = id;
-    particle.capabilities = capabilities;
+
```
Comment (lindner) [inline src/runtime/particle-execution-context.ts]: Please advise on what should be done here..

I tried to move the capabilities into the constructor but then saw permission errors in the console.

Comment (shans) [reply src/runtime/particle-execution-context.ts]: What are the permission errors you saw? I would have thought moving capabilities into the constructor was the right thing to do too.

**Comment by lindner** [comment]
I found out what was going on.

Particle has capabilities as a optional in it's constructor, but all of the subclasses don't offer it on their own constructors.

So I made the param required, changed the subclasses and also found a sneaky mixin that was dropping the capabilities arg.

Adding @sjmiles for review.

**Comment by lindner** [comment]
What's the threat model if a particle modifies capabilities in their own constructor?

File: src/runtime/particle-execution-context.ts (hunk 1 above)
Comment (lindner) [reply src/runtime/particle-execution-context.ts]: [ooops, see below for detailed description]

**Comment by sjmiles** [comment]
> What's the threat model if a particle modifies capabilities in their own constructor?

This is a good question. I'm also inclined to **suggest that there be no constructor arguments** simply so user sub-classes don't have to get this right.

File: src/runtime/dom-particle-base.ts [hunk 2]
```diff
@@ -21,9 +21,10 @@ import {Particle} from './particle.js';
 export class DomParticleBase extends Particle {
   private currentSlotName: string | undefined;
 
-  constructor() {
-    super();
+  constructor(capabilities?: {constructInnerArc?: Function}) {
```
[the remark this answers is not in the document]
Comment (lindner) [reply src/runtime/dom-particle-base.ts]: mistake on my part.

But maybe we should move this off the constructor...

File: src/runtime/particle-execution-context.ts [hunk 3]
```diff
@@ -195,19 +195,20 @@ export class ParticleExecutionContext {
...
-  async _instantiateParticle(id: string, spec: ParticleSpec, proxies: ReadonlyMap<string, StorageProxy>) {
+  // tslint:disable-next-line: no-any
+  async _instantiateParticle(id: string, spec: ParticleSpec, proxies: ReadonlyMap<string, StorageProxy>): Promise<[any, () => Promise<void>]> {
     let resolve : () => void = null;
     const p = new Promise<void>(res => resolve = res);
     this.pendingLoads.push(p);
     const clazz = await this.loader.loadParticleClass(spec);
     const capabilities = this.defaultCapabilitySet();
-    const particle = new clazz(); // TODO: how can i add an argument to DomParticle ctor?
-    particle.id = id;
```
[the remark this answers is not in the document]
Comment (lindner) [reply src/runtime/particle-execution-context.ts]: there is no id property on particle.

File: src/runtime/loader.ts [hunk 4]
```diff
@@ -16,6 +16,7 @@ import {JsonldToManifest} from './converters/jsonldToManifest.js';
 import {DomParticle} from './dom-particle.js';
```
[the remark this answers is not in the document]
Comment (lindner) [reply src/runtime/loader.ts]: looks like capabilities are not used there.

**Comment by lindner** [comment]
Rewrote this as a write-once set method.

File: src/runtime/test/artifacts/transformations/test-multiplex-slots-particle.js [hunk 5]
```diff
@@ -8,8 +8,8 @@
 
 defineParticle(({TransformationDomParticle}) => {
   return class MultiplexSlotsParticle extends TransformationDomParticle {
-    constructor() {
-      super();
+    constructor(capabilities) {
```
Comment (sjmiles) [inline src/runtime/test/artifacts/transformations/test-multiplex-slots-particle.js]: Should this be reverted?

Comment (lindner) [reply src/runtime/test/artifacts/transformations/test-multiplex-slots-particle.js]: Thank you.  Done.

**Pull request merged**